docs/data-retention-mechanisms.md

```diff
@@ -0,0 +1,65 @@
+# Data retention mechanisms
+
+There are several mechanisms regarding the retention of data in EmailAlertApi.
+This doc gives a brief overview of each.
+
+## Deactivating subscribers
+
+We deactivate subscribers for a couple of different scenarios. One being where
+the user unsubscribes via either following an unsubscribe link in the email
+or via the manage subscriptions page.
+
+Another way a user is deactivated is if their subscription is ended on our side
+because the topic they're signed up for no longer exists. This happens via an
+unpublished rabbitMQ message which is eventually handled by the
+[UnpublishHandleService]. This is only applicable for subscriptions to the
+topic taxonomy.
+
+[UnpublishHandleService]: https://github.com/alphagov/email-alert-api/blob/master/app/services/unpublish_handler_service.rb
+
+## Nullifying deactivated subscribers
+
+We nullify the email addresses of subscribers after they've been deactivated
+for over 28 days. This means that the address field on the subscriber record is
+updated to nil once the threshold has been reached.
+
+We use the [NullifyDeactivatedSubscribersWorker] to do this which runs [every
+hour].
+
+[every hour]: https://github.com/alphagov/email-alert-api/blob/master/config/sidekiq.yml#L26-L28
+[NullifyDeactivatedSubscribersWorker]: https://github.com/alphagov/email-alert-api/blob/master/app/workers/nullify_deactivated_subscribers_worker.rb
+
+## Archiving emails
+
+We archive emails to S3 using the [EmailArchiveWorker]. Emails are eligible for
+archiving immediately after creation as long as they haven't already been
+archived (based on `archived_at` value).
+
+The [archive worker runs every hour].
+
+[EmailArchiveWorker]: https://github.com/alphagov/email-alert-api/blob/master/app/workers/email_archive_worker.rb
+[archive worker runs every hour]: https://github.com/alphagov/email-alert-api/blob/ec075c7198b965d3771726946dcd37783b9de9c5/config/sidekiq.yml#L29-L31
+
+## Deleting emails
+
+We delete emails after they've been archived for a minimum of 7 days. Emails get
+archived immediately (as soon as they're generated) and then deleted using the
+[EmailDeletionWorker].
+
+The [deletion worker runs every hour].
```

Comment:
A future reader may be confused about why we have a specific worker for deleting emails, and why it runs hourly. Can we add more detail here?

Reply:
Sure, added more detail!

Reply:
Thanks. I actually wonder if we could combine the `EmailDeletionWorker` and the `HistoricalDataDeletionWorker` together, if the latter runs more frequently. Is there a reason we can't do that?

That's what I think we should explain here. It may be there is no reason and it's purely historical. If so, we could say we're planning to combine them as part of removing the archiving feature (as an idea).

Reply:
Theres no reason why we can't combine the workers, technically speaking. However when the historic worker was created it was decided that the deletion of emails was solving a separate issue to what the historic worker was intended for, hence the different scheduled worker run times.

E.g: 

* Email deletion worker - to keep the db performant (as v large quantity of emails produced)

* Historical deletion worker - keeping the database tidy and manageable (not perpetually keeping unused data)

I think both workers responsibilities are now converging though so it may make sense to combine them. An example being is that we may have to start deleting digest_run_subscribers sooner and sooner as we produce more and more and at that point the historic worker will have v similar responsibilities to the email deletion worker.

I'll add a bit in to explain that's what we're planning to do when retiring archiving. I guess this means I'll have to update the ADR too.

Reply:
Thanks @1pretz1, a caveat in this section sounds good. Once archiving is gone I agree it makes sense to look at combining them, as long as no one part of the deletion takes a long time (on its own).

I don't think you need to change the ADR: the policy is unchanged; we're just talking about the location of the code, which is always up for change in the future. ADRs don't need to be permanently accurate.